src/multisig/multisig_kex.cpp

```diff
@@ -0,0 +1,638 @@
+      //  with them for the next round
+      else
+        msg_pubkeys.push_back(derivation);
+    }
+
+    return multisig_kex_msg{round, base_privkey, msg_pubkeys};
+  }
+  //----------------------------------------------------------------------------------------------------------------------
+  /**
+   * INTERNAL
+   *
+   * brief: multisig_kex_msgs_sanitize_pubkeys - Sanitize multisig kex messages.
+   *    - Removes duplicates from msg pubkeys, ignores pubkeys equal to the local account's signing key,
+   *      ignores messages signed by the local account, ignores keys found in input 'exclusion set',
+   *      constructs map of pubkey:origins.
+   *    - Requires that all input msgs have the same round number.
+   *
+   *    origins = all the signing pubkeys that recommended a given pubkey found in input msgs
+   *
+   *    - If the messages' round numbers are all '1', then only the message signing pubkey is considered
+   *      'recommended'. Furthermore, the 'exclusion set' is ignored.
+   * param: own_pubkey - local account's signing key (key used to sign multisig messages)
+   * param: expanded_msgs - set of multisig kex messages to process
+   * param: exclude_pubkeys - pubkeys to exclude from output set
+   * outparam: sanitized_pubkeys_out - processed pubkeys obtained from msgs, mapped to their origins
+   * return: round number shared by all input msgs
+   */
+  //----------------------------------------------------------------------------------------------------------------------
+  static std::uint32_t multisig_kex_msgs_sanitize_pubkeys(const crypto::public_key &own_pubkey,
+    const std::vector<multisig_kex_msg> &expanded_msgs,
+    const std::vector<crypto::public_key> &exclude_pubkeys,
+    std::unordered_map<crypto::public_key_memsafe, std::unordered_set<crypto::public_key>> &sanitized_pubkeys_out)
+  {
+    CHECK_AND_ASSERT_THROW_MES(expanded_msgs.size() > 0, "At least one input message expected.");
+
+    std::uint32_t round = expanded_msgs[0].get_round();
+    sanitized_pubkeys_out.clear();
+
+    // get all pubkeys from input messages, add them to pubkey:origins map
+    // - origins = all the signing pubkeys that recommended a given msg pubkey
+    for (const auto &expanded_msg : expanded_msgs)
+    {
+      CHECK_AND_ASSERT_THROW_MES(expanded_msg.get_round() == round, "All messages must have the same kex round number.");
+
+      // ignore messages from self
+      if (expanded_msg.get_signing_pubkey() == own_pubkey)
+        continue;
+
+      // in round 1, only the signing pubkey is treated as a msg pubkey
+      if (round == 1)
+      {
+        // note: ignores duplicates
+        sanitized_pubkeys_out[expanded_msg.get_signing_pubkey()].insert(expanded_msg.get_signing_pubkey());
+      }
+      // in other rounds, only the msg pubkeys are treated as msg pubkeys
+      else
+      {
+        // copy all pubkeys from message into list
```

Comment:
I tried to make the messaging protocol flexible. In this case, duplicate input keys are permitted (they are ignored by the `unordered_map` and `unordered_set` use in `evaluate_multisig_kex_round_msgs()`).

Reply:
There aren't checks to: (1) ensure the remote party produced a `public_key` identical to the one produced locally, (2) that all of the messages came from the same signers, and (3) a check to ensure the aggregate pubkey is identical across all parties. Or at least, I don't recall such checks. Of all of them, (1) requires the least amount of code (now) and provides some additional sanity checks against incorrect messaging.

Reply:
This function is 'sanitize pubkeys', it is only responsible for filtering data out of the input messages. The 'evaluate msgs' function does all these kinds of data checks.

> (1) ensure the remote party produced a `public_key` identical to the one produced locally

I explicitly don't do this test so 'round boosting' can work seamlessly (future PR). I assume you are referring to intermediate kex keys, not the final aggregate key.

Why isn't it a problem?
- If there are more rounds left, then we don't care if remote parties have the same keys as us. We take the messages from this round, remove all keys that match local keys, then make DH secrets for the next round. In the next round, we _don't care_ about our keys from the previous round.
  - If the message we make is for the last round, then we save the DH secrets as our final secret keys (who cares if other participants have these keys or not? if they lost their keys, then I guess the local signer just has more signing power than expected).
- If we are in the last round (we will make the final aggregate key from the set of messages at hand), then we already have all our private keys corresponding to signer-sets of size `N - M + 1`. All we care about is if _every set of size `N - M + 1` that excludes us shares a key_. If that condition holds, then if any set of `N - M` signers (other than the local signer) are removed, then there is always at least one person remaining with each key piece. That is the basic requirement of multisig: remove `N - M` signers, and the remaining `M` signers can always complete a signature (note: it is nonsensical from a security p.o.v. for the local signer to validate whether removing himself will allow the other `M` signers to complete a signature).

> (2) that all of the messages came from the same signers

This check is performed in `evaluate_multisig_kex_round_msgs()` where I comment `// 2 (continued). only expected signers should be recommending keys`.

> (3) a check to ensure the aggregate pubkey is identical across all parties

If there are fewer than `N - M + 1` malicious signers, then identical aggregate pubkeys between honest signers are implied, because each signer signs their kex messages. If there are `N - M + 1` malicious signers, then validating the aggregate pubkey is futile because your setup is compromised.

Reply:
I was also concerned about unintentional glitches when computing the shared secrets since there is no direct lookups to see that remote and local computations match. However, this would require that two shared secrets from one signer _accidentally_ were equivalent public keys, which is extremely unlikely. Probably about as likely as two people randomly selecting the same private key.

As you pointed out, the malicious attacker scenario isn't worth covering, because a malicious co-signer can just refuse to sign, etc., so there's no point in intentionally subverting this process.

And to be clear (I think you understand what